### PR TITLE
feat:where filter

### LINKED ISSUE #371: Enable filters on meta data
### 🚀 The feature

- https://discord.com/channels/1121119078191480945/1125758905310519327/1133469950196064387
- https://discord.com/channels/1121119078191480945/1125758905310519327/1133811371339944117

### Motivation, pitch

better filtering.

### Alternatives

_No response_

### Additional context

_No response_

### LINKED ISSUE #395: Add support of where filter, so that user can filter documents accordingly.
### 🚀 The feature

Often talked about, but still not part of our code base. 

### Motivation, pitch

`where` filters can be used to filter the database, and give metadata a purpose.

Example uses: 
- user specific embeddings
- date filter
- topic filtering
- source filtering (only YT, only qna pairs etc.).

### Alternatives

An alternative is using different collections, but that means if you have lots of duplicates you will have to store everything multiple times, so it's not a real alternative.

### Additional context

There where filter should be implemented:
* for query
* for chat
* in a new delete method

It should be part of the queryConfig.

### PR DESCRIPTION
**Pull request by sw8fbar**
## Description

Options to add where_filter on the retrieve_from_database function on embedchain.py. The options can be included in the QueryConfig or provided when the .query or .chat functions are called on EmbedChain or its subclasses. 

Fixes #395
Fixes #371

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

```
from string import Template
from embedchain import App
from embedchain.config import ChatConfig, QueryConfig
from langchain.memory import ChatMessageHistory

app = App()
# add data
app.add("test-abcd", 'text', metadata={"type": "alpha"})
app.add("test-23232331121212", 'text', metadata={"type": "num"})

# test for where_filter in query / chat
print({"type": "alpha"}, app.retrieve_from_database("return data", QueryConfig(), where_filter={"type": "alpha"}))
print({"type": "num"}, app.retrieve_from_database("return data", QueryConfig(), where_filter={"type": "num"}))

# test for where_filter in QueryConfig
queryconfig = QueryConfig(where_filter={"type": "alpha"})
print({"type": "alpha"}, app.retrieve_from_database("return context",queryconfig))
queryconfig = QueryConfig(where_filter={"type": "num"})
print({"type": "num"},app.retrieve_from_database("return context",queryconfig))

# test for where_filter in ChatConfig
chatconfig = ChatConfig(where_filter={"type": "alpha"})
chatconfig.set_history(ChatMessageHistory())
print({"type": "alpha"},app.retrieve_from_database("return data", chatconfig))
chatconfig = ChatConfig(where_filter={"type": "num"})
chatconfig.set_history(ChatMessageHistory())
print({"type": "num"}, app.retrieve_from_database("return data", chatconfig))
```

Please delete options that are not relevant.

- [X] Test Script (please provide)

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] I have checked my code and corrected any misspellings

## Maintainer Checklist

- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] Made sure Checks passed

### CODE REVIEW

**Comment by sw8fbar** [comment]
@cachho please review if this meets your needs

**Comment by cachho** [comment]
I think this most certainly needs a unit test

File: embedchain/embedchain.py [hunk 1]
```diff
@@ -247,16 +247,24 @@ def get_llm_model_answer(self):
...
+        :param where_filter: Optional. A dictionary of key-value pairs to filter the database results.
         :return: The content of the document that matched your query.
         """
-        where = {"app_id": self.config.id} if self.config.id is not None else {}  # optional filter
+        if where_filter is not None:
+            where = where_filter
+        elif config is not None and config.where_filter is not None:
+            where = config.where_filter
+        elif self.config.id is not None:
+            where = {"app_id": self.config.id}
```
[the remark this answers is not in the document]
Comment (cachho) [reply embedchain/embedchain.py:265]: note the `if`, it can't be in the same block as the supplied `where` filters

**Review by cachho — CHANGES_REQUESTED** [review]
Other than that it looks good.

File: embedchain/embedchain.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sw8fbar) [reply embedchain/embedchain.py]: Made requested changes. 
`
       if where_filter is not None:
            where = where_filter
        elif config is not None and config.where_filter is not None:
            where = config.where_filter

        if self.config.id is not None:
            where.update({"app_id": self.config.id})
`

**Comment by sw8fbar** [comment]
Made requested changes. 
 ```
 if where_filter is not None:
            where = where_filter
        elif config is not None and config.where_filter is not None:
            where = config.where_filter

        if self.config.id is not None:
            where.update({"app_id": self.config.id})
```

**Review by cachho — APPROVED** [review]
Looks good. Do you think you will add unit tests?

**Comment by sw8fbar** [comment]
I think it should be the following though. I can push this change if you agree

      ```
       if where_filter is not None:
           where = where_filter
       elif config is not None and config.where_filter is not None:
           where = config.where_filter
       else:
           where = {}

       if self.config.id is not None:
           where.update({"app_id": self.config.id})
```

**Comment by cachho** [comment]
> I am working on it. Just struggling a bit to get it to work. My approach was to test if the parameters are passed to the retrieve_from_database function. But I am not sure that is enough.

I think that will do for now. I think it's technically possible to just insert vectors with metadata, and then you should also be able to search with a vector, and see if the result is what you expect. But I don't know how well Chroma exposes this stuff, and we don't want to write a test suite for chroma.

**Comment by sw8fbar** [comment]
> > I am working on it. Just struggling a bit to get it to work. My approach was to test if the parameters are passed to the retrieve_from_database function. But I am not sure that is enough.
> 
> I think that will do for now. I think it's technically possible to just insert vectors with metadata, and then you should also be able to search with a vector, and see if the result is what you expect. But I don't know how well Chroma exposes this stuff, and we don't want to write a test suite for chroma.

Added the Unit tests

**Review by taranjeet — APPROVED** [review]

File: embedchain/apps/PersonApp.py [hunk 2]
```diff
@@ -63,9 +62,9 @@ def query(self, input_query, config: QueryConfig = None, dry_run=False):
         config = self.add_person_template_to_config(DEFAULT_PROMPT, config)
         return super().query(input_query, config, dry_run)
 
-    def chat(self, input_query, config: ChatConfig = None, dry_run=False):
+    def chat(self, input_query, config: ChatConfig = None, dry_run=False, where_filter=None):
```
[the remark this answers is not in the document]
Comment (sw8fbar) [reply embedchain/apps/PersonApp.py]: > Hey @sw8fbar , can you please update the name to `where` instead of `where_filter`?

Done.

**Comment by taranjeet** [comment]
The PR looks good. Thanks @sw8fbar for your contribution.

**Comment by taranjeet** [comment]
@sw8fbar : Do you want to create a new PR where you add the documentation for this change as well?

**Comment by sw8fbar** [comment]
> @sw8fbar : Do you want to create a new PR where you add the documentation for this change as well?

Sure. Give me a few

**Comment by sw8fbar** [comment]
> > @sw8fbar : Do you want to create a new PR where you add the documentation for this change as well?
> 
> Sure. Give me a few

Added pull request #547

**Pull request merged**